Consensys/linea-token-list
# Czk 458

This pull request introduces a GitHub Action workflow that automates the generation of the `linea-mainnet-token-fulllist.json` file. This JSON file contains information about various tokens and is updated hourly.

### Changes Made:

- Added a GitHub Action workflow definition in the .github/workflows directory. The workflow is configured to run every hour.

- Created a TypeScript script (createOnChainTokenList.ts) that fetches token information from Ethereum contracts, processes it, and generates the linea-mainnet-token-fulllist.json file.

- Configured the GitHub Action workflow to run the TypeScript script and commit changes to the repository only if there are modifications.


### How to Test This PR:

To test this pull request, try locally by following these steps:


To try the changes from your PR locally, you can follow these steps:

1. Create a .env file by copying the contents of .env.example and replace PROVIDER_URL and LINEA_PROVIDER_URL by the correct values.

2. Replace the content of the `linea-mainnet-token-fulllist.json` file with the following empty token list structure:
```{
  "type": "LineaTokenList",
  "tokenListId": "https://raw.githubusercontent.com/Consensys/linea-token-list/main/json/linea-mainnet-token-fulllist.json",
  "name": "Linea Mainnet Token Full List",
  "createdAt": "2023-09-13",
  "updatedAt": "2023-09-18",
  "versions": [
    {
      "major": 1,
      "minor": 0,
      "patch": 0
    }
  ],
  "tokens": []
}
```

3. Open your terminal and navigate to the project directory.

4. Run `yarn install` to install the required Node.js modules.

5. Run `yarn start` to execute the script (`createOnChainTokenList.ts`). The script will fetch token information, process it, and generate an updated `linea-mainnet-token-fulllist.json` file in your project directory.

6. Check the generated `linea-mainnet-token-fulllist.json` file to ensure it contains the expected token list structure.

### Informations about the Token List:

The Token List is a structured JSON file that contains information about various tokens, including their metadata. Here's how the Token List is generated and updated based on the provided script:

- Fetching Token Information: The script listens to two types of events, NewToken and NewTokenDeployed, to identify new tokens or tokens deployed on a different chain. Then the script fetches additional informations directly from the ERC-20 token contracts themselves.

- NewToken Event: This event represents the creation of a new token. The chainId, chainURI, and other fields are set based on the event type. For example, the chainId will be `59144` in this case.
 
- NewTokenDeployed Event: This event represents the deployment of an existing token to a different chain. Similar to the `NewToken` event, it sets the chainId, chainURI, and other fields accordingly.

- Fetching Logo URI: The script attempts to fetch the token's logo URI from CoinGecko using the token's address. If successful, it adds the logoURI field to the token's information. The logo URI is used for displaying the token's logo in user interfaces.
 
- Versioning: If the script detects a new token (either NewToken or NewTokenDeployed event), it increments the minor version in the Token List. This versioning mechanism allows users and applications to track changes and updates to the Token List.

- Updating the Token List: The script updates the Token List by adding new tokens to the list. It checks if the token's address is already in the existing Token List to avoid duplicates. New tokens are added to the list, and the list is saved with the updated information.

In summary, the script fetches token information from specific events, including the token's metadata, chain information, and logo URI. It then updates the Token List with this information, increments the version if necessary, and ensures that the Token List remains up to date with the latest tokens and their details.
